{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 7755.34x400 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#This code was made by Hideo NAKANO to map epitope along with the protein sequence of target \n",
    "#based on enrithment ration of three amino acid combination calculated from NGS analysis\n",
    "# with the help of ChatGPT 20230725\n",
    "# modified to remove the surrounding amino acids\n",
    "# changed to 2 aa combinations 20230725\n",
    "\n",
    "import pandas as pd  # Data manipulation and analysis\n",
    "import matplotlib.pyplot as plt  # Plotting library\n",
    "import math  # Mathematical functions\n",
    "from Bio import SeqIO  # For reading and writing sequence file formats\n",
    "import csv  # For working with CSV files\n",
    "\n",
    "\n",
    "# Function to calculate ratio\n",
    "def calc_ratio(pep, ngs_sum_list, count_sum):\n",
    "    # Initialize counters\n",
    "    count_a = 0  \n",
    "    count_p = 0  \n",
    "    # Iterate over each item in list\n",
    "    for item in ngs_sum_list:\n",
    "        # If the peptide is found in the item's 2nd position\n",
    "        if pep in item[1]:  \n",
    "            count_p += item[2]  # Increment peptide count\n",
    "            count_a += 1  # Increment count\n",
    "    # Calculate ratio\n",
    "    count_r = count_p/count_sum  \n",
    "    # Return a list with peptide, ratio and count\n",
    "    return [pep, count_r, count_a]  \n",
    "\n",
    "# Function to enrich peptide\n",
    "def enrich_peptide(pep_list_list, ngs_sum_list_ori, ngs_sum_list_r1, count_sum_ori, count_sum_r1):\n",
    "    # Initialize output lists\n",
    "    out_ori = [[\"peptide\", \"ratio_sum_ori\", \"count_ori\"]]  \n",
    "    # Extend the list with the calculated ratios\n",
    "    out_ori.extend(list(map(lambda pep: calc_ratio(pep, ngs_sum_list_ori, count_sum_ori), pep_list_list)))  \n",
    "\n",
    "    # Similar steps for another output list\n",
    "    out_r1 = [[\"peptide\", \"ratio_sum_r1\", \"count_ori\"]]  \n",
    "    out_r1.extend(list(map(lambda pep: calc_ratio(pep, ngs_sum_list_r1, count_sum_r1), pep_list_list)))  \n",
    "\n",
    "    # Final output list\n",
    "    out=[[\"peptide\", \"EF\", \"ratio_ori\", \"ratio_r1\"]]  \n",
    "    for i in range(1,len(out_ori)):\n",
    "        if out_ori[i][1] > 0:  # Check if the ratio is greater than 0\n",
    "            EF = out_r1[i][1]/out_ori[i][1]  # Calculate enrichment factor\n",
    "            out.append([out_ori[i][0],EF,out_ori[i][1],out_r1[i][1]])  # Append to the output list\n",
    "        else:\n",
    "            out.append([out_ori[i][0],'NaN',out_ori[i][1],out_r1[i][1]])  # Append 'NaN' if ratio is 0\n",
    "    return out  # Return the output list\n",
    "\n",
    "# Main function where the program starts\n",
    "def main():\n",
    "    # Specify names of the CSV files and plot\n",
    "    csv_name = \"epitope_out_400_R1.csv\"  #output file of enrichment factor of  three amino acid epitope \n",
    "    pep_list_name = \"ep_pep_list400.csv\" #input file of amino acid combination list \n",
    "    ngs_sum_ori_name = \"NostarX_Round0.csv\" # Summary of the initial NGS data o\n",
    "    ngs_sum_r1_name = \"NostarX_Round3.csv\"  #Summnary of the NGS data after selection\n",
    "    fasta_target = list(SeqIO.read(\"HA-NEW.fa\", \"fasta\"))  # Read target fasta sequence.\n",
    "    csv_name_target= \"antiHAtag_R1_2aa.csv\"  #CSV output of the target seq\n",
    "    plot_name_target = \"antiHAtag_R1_2aa.jpeg\" # Plot file of the target  \n",
    "\n",
    "    # Load data from CSV files into pandas DataFrames\n",
    "    ngs_sum_ori = pd.read_csv(ngs_sum_ori_name, keep_default_na=False, na_values=['_'])   \n",
    "    ngs_sum_r1 = pd.read_csv(ngs_sum_r1_name, keep_default_na=False, na_values=['_'])\n",
    "    pep_list = pd.read_csv(pep_list_name, keep_default_na=False, na_values=['_'])\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    ngs_sum_ori['pep'] =  ngs_sum_ori.iloc[:,1]  \n",
    "    ngs_sum_r1['pep'] =  ngs_sum_r1.iloc[:,1] \n",
    "\n",
    "    # Calculate the sum of column 2 in both DataFrames\n",
    "    count_sum_ori = ngs_sum_ori.iloc[:,2].sum()  \n",
    "    count_sum_r1 = ngs_sum_r1.iloc[:,2].sum()  \n",
    "\n",
    "    # Convert the data in DataFrame to list format\n",
    "    pep_list_list = pep_list.values.flatten().tolist()  \n",
    "    ngs_sum_list_ori = ngs_sum_ori.values.tolist()  \n",
    "    ngs_sum_list_r1 = ngs_sum_r1.values.tolist()  \n",
    "\n",
    "    # Call the function enrich_peptide to calculate enrichment\n",
    "    out = enrich_peptide(pep_list_list, ngs_sum_list_ori, ngs_sum_list_r1, count_sum_ori, count_sum_r1)  \n",
    "\n",
    "    # Write the output to a CSV file\n",
    "    with open(csv_name, \"w\", newline=\"\") as f:  \n",
    "        writer = csv.writer(f)  \n",
    "        writer.writerows(out)  \n",
    "\n",
    "    # Load the CSV file into a DataFrame\n",
    "    pep_enrich = pd.read_csv(csv_name, keep_default_na=False, na_values=['_'])  \n",
    "\n",
    "    # Read a fasta sequence\n",
    "    \n",
    "    # Break the sequence into sub-sequences of length 2\n",
    "    tar_seq = [''.join(fasta_target[i:i+2]) for i in range(len(fasta_target)-1)]  \n",
    "\n",
    "    # Create a new DataFrame\n",
    "    df = pd.DataFrame({'Seq': fasta_target, 'EF': [0]*len(fasta_target)})  \n",
    "\n",
    "    # Update the 'EF' column in DataFrame based on 'peptide' column in another DataFrame\n",
    "    for i, item in enumerate(tar_seq):\n",
    "        if i>=len(tar_seq):\n",
    "            break\n",
    "        df.loc[i:i+1, 'EF'] += pep_enrich.loc[pep_enrich['peptide'].str.contains(item), 'EF'].sum()  \n",
    "\n",
    "    # Save the DataFrame to a CSV file\n",
    "    df.to_csv(csv_name_target)  \n",
    "\n",
    "    # Calculate length for the plot\n",
    "    x_len = math.sqrt(len(fasta_target))*4+5  \n",
    "    plt.rcParams[\"figure.figsize\"] = [x_len,4.0]  \n",
    "    # Plot the data\n",
    "    df.plot(kind='bar',x='Seq', width=1)  \n",
    "    plt.xticks(rotation=0)  \n",
    "    plt.savefig(plot_name_target)  # Save the plot as a JPEG file\n",
    "    plt.show()  # Display the plot\n",
    "\n",
    "# Run the main function\n",
    "if __name__ == \"__main__\":  \n",
    "    main()  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
